### PR TITLE
Parallelize constraint evaluation

### PR DESCRIPTION
**Pull request by gabrielbosio**
I've run the benchmarks with the rayon threadpool configured to use 8 threads on:
- An Apple M1 with 4 E and 4 P cores. With this configuration, this PR improves the performance of the prover by 26% for fibo 10 and 30% for fibo 100.
- An Intel Xeon Platinum with 4 cores. With this configuration, this PR improves the performance of the prover by 14% for fibo 10 and 19% for fibo 100.

### CODE REVIEW

File: src/starks/constraints/evaluator.rs [hunk 1]
```diff
@@ -5,7 +5,12 @@ use lambdaworks_math::{
     traits::ByteConversion,
 };
 
-#[cfg(debug_assertions)]
+#[cfg(feature = "parallel")]
+use rayon::prelude::{
+    IndexedParallelIterator, IntoParallelIterator, IntoParallelRefIterator, ParallelIterator,
+};
+
+#[cfg(all(debug_assertions, not(feature = "rayon")))]
```
Comment (entropidelic) [inline src/starks/constraints/evaluator.rs:13]: seems like it should be `parallel` instead of `rayon` :) 

also check the others

**Review by diegokingston — APPROVED** [review]

File: src/starks/constraints/evaluator.rs [hunk 2]
```diff
@@ -297,3 +313,45 @@ impl<'poly, F: IsFFTField, A: AIR + AIR<Field = F>> ConstraintEvaluator<'poly, F
...
+            evaluate_polynomial_on_lde_domain(
+                exemption,
+                domain.blowup_factor,
+                domain.interpolation_domain_size,
+                &domain.coset_offset,
+            )
+            .unwrap()
+        })
+        .collect()
+}
```
Comment (MauroToscano) [inline src/starks/constraints/evaluator.rs:357]: Does it work if you just #[cfg(feature = "parallel")] the .par_iter() ?

Comment (MauroToscano) [reply src/starks/constraints/evaluator.rs:357]: Same in other places

**Review by MauroToscano — APPROVED** [review]

**Comment by diegokingston** [comment]
In future version, we should only compute the necessary exemptions and then change the way the transition polynomials are computed

**Review by entropidelic — APPROVED** [review]

**Pull request merged**